cvc5/Init.lean

```diff
@@ -0,0 +1,193 @@
+      )
+
+  if let (some autoDoc?, some autoId) := (autoDoc, autoId) then
+    let env ← getEnv
+    let id := ident.getId.toString
+    let name := Lean.Name.mkSimple id
+    let ns ← Elab.liftMacroM Macro.getCurrNamespace
+    let name := ns.append name
+    let arity ←
+      if let some (.opaqueInfo i) := env.find? name then
+        pure (forallArity i.type)
+      else
+        throwError s!"failed to retrieve arity of (opaque) function `{ident}`"
+
+    let mut args := Array.empty
+    for i in [0:arity] do
+      let arg := Lean.Name.mkSimple s!"v{i}" |> Lean.mkIdent
+      args := args.push arg
+    let funCall : TSyntax `term ← `(term| ($ident $[ $args ]* ))
+
+    for (autoDoc?, autoId) in autoDoc?.zip autoId do
+      let id : String := autoId.getId.toString
+      let body ←
+        if id.endsWith "!" then
+          let unwrapName := Lean.Name.mkStr3 "cvc5" "Error" "unwrap!" |> Lean.mkIdent
+          `(fun $[$args]* => $funCall |> $unwrapName)
+        else if id.endsWith "?" then
+          `(fun $[$args]* => $funCall |> Except.toOption)
+        else
+          throwError s!"unexpected auto function name `{id}`: expected `<ident>!` or `<ident>?`"
+      Command.elabCommand
+        (← define autoDoc? autoId (← `(optDeclSig|)) body)
+
+  if let
+    (some subDoc?, some subId, some subSig, some subDef)
+    := (subDoc, subId, subSig, subDef)
+  then
+    let all := subDoc?.zip subId |>.zip subSig |>.zip subDef
+    for (((subDoc?, subId), subSig), subDef) in all do
+      Command.elabCommand
+        (← define subDoc? subId subSig subDef)
+| _ => throwUnsupportedSyntax
+
+/-- Defines similar functions realized by `extern`.
+
+```
+extern! "prefix"
+  /-- Create a Boolean constant.
+
+  - `b`: The Boolean constant.
+
+  Will create an opaque definition with `[@extern extStr]` where
+  `extStr = "prefix" ++ "_" ++ "myFunction"`.
+  -/
+  def myFunction : Term → Except Error Op
+  with
+    endsWithBang!
+    endWithQuestion?
+  where
+    myOtherFunction : Term → Op :=
+      Error.unwrap! ∘ myFunction
+    /-- Optional function docstring: if none, inherit from the main function. -/
+    yetAnotherFunction : Term → Option Op :=
+      Except.toOption ∘ myFunction
+```
+
+- `with ...`: takes a sequence of identifiers, each generate a function that
+  - unwraps the result if `!`-ended, which generates code similar to `myOtherFunction` above;
+  - turns a result into an option if `?`-ended, which generates code similar to `yetAnotherFunction`
+    above;
+  - fails otherwise.
+
+  The `with ...` syntax is currently only compatible with external functions that produce `Except
+  Error α` values.
+
+- Supports `declModifiers` on the main (`def`) function `myFunction` such as `private`.
+- Accepts a list of external (`def`) functions, each with its `with` and/or `where` clauses.
+-/
+scoped syntax (name := multidefs)
+  withPosition("extern! " str ppLine group(colGt defsItem)+)
+: command
```

Comment:
Are you using this command to avoid listing the namespace/prefix for function names each time? If not, I would prefer not to use it, as I prefer each function to be completely separate from the others. I am fine with something like `extern_def` or `cvc5_def`.

Reply:
I assume your question is about the macro supporting more than one definition. If that's the case, then the point was to avoid writing `extern! <pref>` dozens of times for each definition separately since this is a DRY macro.

Let me know if you want to have one `extern! <pref>` per definition.

Reply:
I don't want to mention `<pref>` at all. Is there a way to automatically detect it? If not, then it's fine to use `extern!`. Otherwise, I would prefer to not use this macro.

> Is there a way to automatically detect it?

Note: this may not be possible with a pure macro...